### ISSUE TITLE
[BUG] [csharp] UserAgent not updated in CallApiAsync

### ISSUE THREAD

**Issue by jamesrusso** [OpenAPITools/openapi-generator#3595]
In ApiClient.cs, The UserAgent is not updated in the CallApiAync method, but it in the CallApi method.

```
      public Object CallApi(
            String path, RestSharp.Method method, List<KeyValuePair<String, String>> queryParams, Object postBody,
            Dictionary<String, String> headerParams, Dictionary<String, String> formParams,
            Dictionary<String, FileParameter> fileParams, Dictionary<String, String> pathParams,
            String contentType)
        {
            var request = PrepareRequest(
                path, method, queryParams, postBody, headerParams, formParams, fileParams,
                pathParams, contentType);

            // set timeout
            
            RestClient.Timeout = Configuration.Timeout;
            // set user agent
           RestClient.UserAgent = Configuration.UserAgent;

            InterceptRequest(request);
            var response = RestClient.Execute(request);
            InterceptResponse(request, response);

            return (Object) response;
        }
```

vs

```
        public async System.Threading.Tasks.Task<Object> CallApiAsync(
            String path, RestSharp.Method method, List<KeyValuePair<String, String>> queryParams, Object postBody,
            Dictionary<String, String> headerParams, Dictionary<String, String> formParams,
            Dictionary<String, FileParameter> fileParams, Dictionary<String, String> pathParams,
            String contentType)
        {
            var request = PrepareRequest(
                path, method, queryParams, postBody, headerParams, formParams, fileParams,
                pathParams, contentType);
            InterceptRequest(request);
            var response = await RestClient.ExecuteTaskAsync(request);
            InterceptResponse(request, response);
            return (Object)response;
        }
```

I can submit the pull request to fix this template, but wanted some discussion if this should be fixed/set on each request or globally when the RestClient is created. I haven't dug into it this much.

**Comment by jamesrusso** [comment]
```
    public ApiClient(String basePath = "https://xxx")
        {
           if (String.IsNullOrEmpty(basePath))
                throw new ArgumentException("basePath cannot be empty");

            RestClient = new RestClient(basePath);
            Configuration = Client.Configuration.Default;
            RestClient.UserAgent = Configuration.UserAgent;
        }

```

I had thought about putting it in here, where it makes sense, but for some reason Client.Configuration.Default returns null when I do this? Can't figure out exactly why yet. Maybe Something to do with the static constructor not firing on Configuration() class.

**Comment by jamesrusso** [comment]
I don't fully understand it, but the Static Constructor in Configuration object sets Timeout, which then instantiates a new ApiClient, which then calls Configuration again. Since this all originates from the Static constructor, the call to Configuration.Default returns null in initial ApiClient.

**Comment by wing328** [comment]
@jamesrusso to help us understand the issue more, which version of openapi generator are you using? we just release v4.1.0. Please give it a try.

Which C# client generator are you using? `csharp` or `csharp-netcore`?

**Comment by jamesrusso** [comment]
Thanks @wing328. 

Sorry I was not clear. This issue exists on master. This is on the csharp generator. 

The issue originated that UserAgent was not being set. I then forked to work on a PR to add it to the constructor of ApiClient, but the interaction between the static constructor, the static property and the side effect of the timeout properly creating the ApiClient (which also calls Configuration.Default) makes it so the configuration isn’t available in the first instantiation of ApiClient. This results in a NullPointer exception if you try and reference the Configuration object in ApiClient on the first instantiation where the _globalConfiguration is null.

Bottom line is the Static Constructor in Configuration is instantiating a client which is trying to get the Default Configuration but the static variable _globalConfiguration is hasn't yet been set.

**Comment by jamesrusso** [comment]
I've done a PR to fix the underlying issue of UserAgent not being set on Async calls. I still would like to have a discussion about the circular dependency between the Configuration object holding the RestClient which itself creates the Configuration object. 

Basically, if you try and reference the Configuration in the constructor of the ApiClient you'll get a null point exception.
